### PR TITLE
Avoid potentially passing NULL arguments

### PR DESCRIPTION
**Pull request by leorosen**
Several call sites flagged by Coverity that may potentially cause
a pointer argument to be NULL.

In two cases the issue is using a function call as a parameter to
a second function, where the first function may return NULL, while
the second function does not check for the NULL argument value.

Remaining case is when static configuration is mixed with run-time
decision, that could result in a data buffer argument being NULL.

Partially continuing from https://github.com/Mbed-TLS/mbedtls/pull/5700.

### CODE REVIEW

File: library/ssl_tls12_server.c [hunk 1]
```diff
@@ -2854,7 +2855,14 @@ static int ssl_get_ecdh_params_from_cert( mbedtls_ssl_context *ssl )
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
 
-    if( ! mbedtls_pk_can_do( mbedtls_ssl_own_key( ssl ), MBEDTLS_PK_ECKEY ) )
+    const mbedtls_pk_context *private_key = mbedtls_ssl_own_key( ssl );
+    if( private_key == NULL)
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "got no ECDH private key" ) );
```
Comment (gilles-peskine-arm) [inline library/ssl_tls12_server.c:2861]: Rather, "got no server private key" — ECDH or not.

Is it actually possible to reach this case, or is it only an internal error?

File: library/ssl_tls12_server.c [hunk 2]
```diff
@@ -3233,6 +3241,12 @@ static int ssl_prepare_server_key_exchange( mbedtls_ssl_context *ssl,
          */
         if( md_alg != MBEDTLS_MD_NONE )
         {
+            if( dig_signed == NULL )
```
Comment (gilles-peskine-arm) [inline library/ssl_tls12_server.c:3244]: It's too late to check this: if `dig_signed` is null then the initialization of `dig_signed_len` above already invoked undefined behavior.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

**Comment by leorosen** [comment]
Feel free to take over this PR, as I can not continue spend any time on upstreaming our local changes.

File: library/ssl_tls12_server.c (hunk 1 above)
Comment (paul-elliott-arm) [reply library/ssl_tls12_server.c:2861]: I think its possible under a misconfiguration to reach here, but its really only internal.

File: library/ssl_tls12_server.c [hunk 3]
```diff
@@ -3204,6 +3212,14 @@ static int ssl_prepare_server_key_exchange( mbedtls_ssl_context *ssl,
 #endif
         int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
 
+        if( dig_signed == NULL )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
+            return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+        }
```
Comment (gilles-peskine-arm) [inline library/ssl_tls12_server.c:3219]: Please move this check before the declaration of `dig_signed_len`, and initialize it directly to its intended value. We're writing C99 now, and can take advantage of mid-block declarations when no end-of-function cleanup needs to be done.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by daverodgman** [comment]
Probably we should backport?

**Comment by gilles-peskine-arm** [comment]
This is internal robustness, not a bug fix, and the affected code has changed quite a bit, so a backport is not necessary or useful.

**Pull request merged**